### PR TITLE
Feature/qw002

### PR DESCRIPTION
**Pull request by Sar-dim**

### CODE REVIEW

File: QuizWhois.Domain/Services/Implementations/QuestionRatingService.cs [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+            _context.Set<QuestionRating>().Add(entity);
+            _context.SaveChanges();
+            return new QuestionRatingModel(entity.Id, entity.QuestionId, entity.UserId, entity.RatingNumber);
+        }
+
+        public QuestionRatingModel UpdateRating(QuestionModel questionModel, long userId, uint rating)
+        {
+            var entity = _context.Set<QuestionRating>().FirstOrDefault(x => x.QuestionId == questionModel.Id && x.UserId == userId);
+            entity.RatingNumber = rating;
+            _context.Set<QuestionRating>().Update(entity);
```
Comment (Misvor) [inline QuizWhois.Domain/Services/Implementations/QuestionRatingService.cs:52]: If i recall correctly, it's not the way the update in EF Core works. Update only starts tracking entity if it was not tracked before. If you want to update entity fields, just go ahead, then call saveChanges()

File: QuizWhois.Api/Controllers/QuestionRatingController.cs [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+            return Ok(updatingRating);
+        }
+
+        [HttpDelete]
+        [ProducesResponseType(StatusCodes.Status200OK)]
+        [ProducesResponseType(StatusCodes.Status400BadRequest)]
+        public IActionResult Delete(QuestionModel questionModel, long userId)
+        {
+            _questionRatingService.DeleteRating(questionModel, userId);
+            return Ok("Deleted");
```
[the remark this answers is not in the document]
Comment (sidgrouse) [reply QuizWhois.Api/Controllers/QuestionRatingController.cs:47]: good point, though we are not going to process the string. Ok() means that everything is fine, and the string is a kinda mockup. but again, don't stick to the strings, lets just remove it

[the remark this answers is not in the document]
Comment (Sar-dim) [reply QuizWhois.Api/Controllers/QuestionRatingController.cs:47]: Just return Ok(); ?

File: QuizWhois.Api/Controllers/QuestionRatingController.cs [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+using QuizWhois.Domain.Services.Interfaces;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+
+namespace QuizWhois.Api.Controllers
+{
+    [ApiController]
+    [Route("[controller]")]
```
Comment (sidgrouse) [inline QuizWhois.Api/Controllers/QuestionRatingController.cs:13]: does it really work?

File: QuizWhois.Api/Controllers/QuestionRatingController.cs [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+        public IActionResult Post(QuestionModel questionModel, long userId, uint rating)
+        {
+            var addedRating = _questionRatingService.AddRating(questionModel, userId, rating);
+            return Ok(addedRating);
+        }
+
+        [HttpPut]
+        [ProducesResponseType(StatusCodes.Status200OK)]
+        [ProducesResponseType(StatusCodes.Status400BadRequest)]
+        public IActionResult Update(QuestionModel questionModel, long userId, uint rating)
```
Comment (sidgrouse) [inline QuizWhois.Api/Controllers/QuestionRatingController.cs:35]: I am not sure it is needed to change such a small thing as a mark, seems excessive.

File: QuizWhois.Common/Models/QuestionRatingModel.cs [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+using System.Threading.Tasks;
+
+namespace QuizWhois.Common.Models
+{
+    public class QuestionRatingModel
+    {
+        public long Id { get; set; }
+        public long QuestionId { get; set; }
+        public long UserId { get; set; }
+        public uint RatingNumber { get; set; }
```
Comment (sidgrouse) [inline QuizWhois.Common/Models/QuestionRatingModel.cs:14]: Mark? Value? lets rename somehow

File: QuizWhois.Domain/Migrations/20211007085534_quizwhoisdb.cs [hunk 6]
```diff
@@ -0,0 +1,86 @@
+﻿using Microsoft.EntityFrameworkCore.Migrations;
+
+namespace QuizWhois.Domain.Migrations
+{
+    public partial class quizwhoisdb : Migration
```
Comment (sidgrouse) [inline QuizWhois.Domain/Migrations/20211007085534_quizwhoisdb.cs:5]: Let's call classes in PascalCaseConvention. Please rename to Init

File: QuizWhois.Domain/Services/Implementations/QuestionRatingService.cs [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+            _context = context;
+        }
+
+        public double GetAverageRating(long questionId)
+        {
+            if (questionId <= 0)
+            {
+                throw new Exception("Invalid data");
+            }
+            return _context.Set<QuestionRating>().Where(x => x.QuestionId == questionId && x.RatingNumber != 0)
```
Comment (sidgrouse) [inline QuizWhois.Domain/Services/Implementations/QuestionRatingService.cs:28]: why x.RatingNumber != 0 ?

File: QuizWhois.Domain/Services/Implementations/QuestionRatingService.cs [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+                .Select(x => x.RatingNumber).ToList().ConvertAll(x => (double)x).Average();
+        }
+
+        public QuestionRatingModel AddRating(QuestionModel questionModel, long userId, uint rating)
```
Comment (sidgrouse) [inline QuizWhois.Domain/Services/Implementations/QuestionRatingService.cs:32]: Do we really need the whole question model here?

File: QuizWhois.Domain/Services/Implementations/QuestionRatingService.cs [hunk 9]
```diff
@@ -0,0 +1,89 @@
...
+            {
+                throw new Exception("This rating already exist");
+            }
+            var entity = new QuestionRating(questionModel.Id, userId, rating);
+            _context.Set<QuestionRating>().Add(entity);
+            _context.SaveChanges();
+            return new QuestionRatingModel(entity.Id, entity.QuestionId, entity.UserId, entity.RatingNumber);
+        }
+
+        public QuestionRatingModel UpdateRating(QuestionModel questionModel, long userId, uint rating)
```
Comment (sidgrouse) [inline QuizWhois.Domain/Services/Implementations/QuestionRatingService.cs:52]: same

File: QuizWhois.Domain/Services/Implementations/QuestionRatingService.cs [hunk 10]
```diff
@@ -0,0 +1,89 @@
...
+            {
+                throw new Exception("Rating is not found");
+            }
+            entity.RatingNumber = rating;
+            _context.Set<QuestionRating>().Update(entity);
+            _context.SaveChanges();
+            return new QuestionRatingModel(entity.Id, entity.QuestionId, entity.UserId, entity.RatingNumber);
+        }
+
+        public bool DeleteRating(QuestionModel questionModel, long userId)
```
Comment (sidgrouse) [inline QuizWhois.Domain/Services/Implementations/QuestionRatingService.cs:73]: same

Comment (sidgrouse) [reply QuizWhois.Domain/Services/Implementations/QuestionRatingService.cs:73]: lets maybe discuss the possibility of updating and deleting a mark?

**Pull request merged**